[FLINK-1752] Improvements to Kafka Source/Sink & some bugfixes in Flink Streaming

rmetzger opened the pull request:
Hi,
I've spend the last two days on improving the Kafka connectors and the streaming fault tolerance.
Since its the first time I'm touching this code, I would like to get some feedback from the streaming guys (in particular on the changes with the state handling).

Please note that this PR is based on https://github.com/apache/flink/pull/474, so it also contains commits from @gaborhermann.
I'll squash the commits after the first round of feedback.

If you want I can separate the the critical fixes from the Kafka connector rewrite, but i would prefer merging everything at once.

mbalassi commented on flink-java/src/main/java/org/apache/flink/api/java/DataSet.java:
```diff
@@ -182,7 +182,6 @@ protected void fillInType(TypeInformation<T> typeInfo) {
 			ClosureCleaner.clean(f, true);
 		}
-		ClosureCleaner.ensureSerializable(f);
 		return f;
```
Why remove this?

rmetzger replied:
Oh, you're right. I'll add it again.

mbalassi commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/StreamCheckpointCoordinator.scala:
```diff
@@ -75,17 +75,17 @@ class StreamCheckpointCoordinator(val executionGraph: ExecutionGraph,
       context.system.scheduler.schedule(2 * interval,2 * interval,self,CompactAndUpdate)
-      log.debug("[FT-MONITOR] Started Stream State Monitor for job {}{}",
+      log.info("Started Stream State Monitor for job {}{}",
         executionGraph.getJobID,executionGraph.getJobName)
```
+1, this is definitely info level.

mbalassi commented on flink-staging/flink-streaming/flink-streaming-connectors/pom.xml:
```diff
@@ -46,9 +53,13 @@ under the License.
 			<artifactId>kafka_2.10</artifactId>
-			<version>0.8.2.0</version>
+			<version>${kafka.version}</version>
 			<exclusions>
```
Heads-up: The code needed for Kafka 0.8.2.0 is slightly different.

rmetzger replied:
Mh, thats not so easy to solve. We probably have to create different maven modules for the different supported versions (if its not possible to detect and use different methods at runtime)

rmetzger replied:
If this is only about the different names of `leader.getConnectionString()`, we can also play some reflection tricks.

mbalassi commented on flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/SimpleKafkaSource.java:
```diff
@@ -71,7 +67,7 @@ public void cancel() {
 	@Override
 	public void open(Configuration config) throws InterruptedException {
-		initializeConnection();
+//		initializeConnection();
```
Why is this commented out? Eventually we should delete the Kafka connectors using the simple API, there is not much point to keeping them at the current point.

gaborhermann replied:
I agree, no point in keeping SimpleKafkaSource, deleted it.

mbalassi commented on flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/iterator/KafkaOnePartitionIterator.java:
```diff
@@ -41,22 +43,25 @@
  */
-public class KafkaConsumerIterator implements Serializable {
+public class KafkaOnePartitionIterator implements KafkaConsumerIterator, Serializable {
 
```
I'd prefer KafkaSinglePartitionIterator

gaborhermann replied:
+1, changing it.

mbalassi commented on flink-staging/flink-streaming/flink-streaming-connectors/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaITCase.java:
```diff
@@ -0,0 +1,246 @@
+
+	private final int ZK_PORT = 6667;
+	private final int KAFKA_PORT = 6668;
+	private String kafkaHost;
```
These ports maybe already allocated, I'd use NetUtils to get an available one.

gaborhermann replied:
Thanks for the hint, using it.

rmetzger commented:
We're addressed all comments on the pull request.

I'm currently running some tests: https://travis-ci.org/rmetzger/flink/builds/55061414.
I'll merge the change once they've passed.
